### PR TITLE
Commerce 7829

### PR DESCRIPTION
**Pull request by marco-leo**

### CODE REVIEW

**Comment by marco-leo** [comment]
ci:test:sf

**Comment by marco-leo** [comment]
ci:test:relevant

**Comment by marco-leo** [comment]
ci:test:stable

**Comment by marco-leo** [comment]
ci:test:stop

**Comment by marco-leo** [comment]
ci:stop

**Comment by marco-leo** [comment]
ci:test:relevant

**Comment by marco-leo** [comment]
ci:test:stable

File: modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java [hunk 1]
```diff
@@ -419,7 +279,11 @@ public CPInstance addCPInstance(
 			boolean overrideSubscriptionInfo, boolean subscriptionEnabled,
 			int subscriptionLength, String subscriptionType,
 			UnicodeProperties subscriptionTypeSettingsUnicodeProperties,
-			long maxSubscriptionCycles, ServiceContext serviceContext)
+			long maxSubscriptionCycles, boolean deliverySubscriptionEnabled,
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java:282]: is this method to deprecate?

File: modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java [hunk 2]
```diff
@@ -1582,30 +1275,6 @@ public CPInstance updateStatus(long userId, long cpInstanceId, int status)
 		return cpInstancePersistence.update(cpInstance);
 	}
 
-	/**
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java:1585]: do these methods have a replace?

File: modules/apps/commerce/commerce-product-service/service.xml [hunk 3]
```diff
@@ -648,6 +648,10 @@
 		<column db-name="deliverySubTypeSettings" name="deliverySubscriptionTypeSettings" type="String" />
 		<column name="deliveryMaxSubscriptionCycles" type="long" />
 		<column name="unspsc" type="String" />
+		<column name="discontinued" type="boolean" />
+		<column name="discontinuedCPInstanceUuid" type="String" />
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/commerce-product-service/service.xml:652]: Can you explain why we need these new fields?

**Comment by andrea-ale-sbarra** [comment]
I don't understand why we need also to link SKUUUID and CProductId, anyaway why these new fields are not called replaced/replacement SKUUID CPRoductId?

**Comment by marco-leo** [comment]
ci:test:relevant

**Comment by marco-leo** [comment]
ci:test:stable

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-impl/src/main/java/com/liferay/headless/commerce/admin/catalog/internal/util/v1_0/SkuUtil.java [hunk 4]
```diff
@@ -44,6 +45,51 @@ public static CPInstance addOrUpdateCPInstance(
...
+
+				discontinuedCPInstance = cpInstanceService.fetchCPInstance(
+					sku.getDiscontinuedSkuId());
+			}
+
+			if (discontinuedCPInstance != null) {
+				CPDefinition discontinuedCPDefinition =
+					discontinuedCPInstance.getCPDefinition();
+
+				discontinuedCProductId =
```
[the remark this answers is not in the document]
Comment (marco-leo) [reply modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-impl/src/main/java/com/liferay/headless/commerce/admin/catalog/internal/util/v1_0/SkuUtil.java]: updated names has discussed. good catch

File: modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java (hunk 1 above)
Comment (marco-leo) [reply modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java:285]: I wanted to clean a bit this class. Too many overloads

File: modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java [hunk 5]
```diff
@@ -682,7 +388,7 @@ public CPInstance addCPInstance(
 			boolean neverExpire, ServiceContext serviceContext)
 		throws PortalException {
 
-		return cpInstanceLocalService.addCPInstance(
+		return addOrUpdateCPInstance(
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java:391]: use cpInstanceLocalService ref to make this method call

File: modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java [hunk 6]
```diff
@@ -744,10 +409,37 @@ public CPInstance addCPInstance(
...
-		return cpInstanceLocalService.addCPInstance(
+		if (Validator.isBlank(externalReferenceCode)) {
+			externalReferenceCode = null;
+		}
+		else {
+			CPInstance cpInstance = cpInstancePersistence.fetchByC_ERC(
+				serviceContext.getCompanyId(), externalReferenceCode);
+
+			if (cpInstance != null) {
+				return updateCPInstance(
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java:427]: use cpInstanceLocalService ref to make this method call

File: modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java [hunk 7]
```diff
@@ -744,10 +409,37 @@ public CPInstance addCPInstance(
...
+					expirationDateDay, expirationDateYear, expirationDateHour,
+					expirationDateMinute, neverExpire, unspsc, discontinued,
+					replacementCPInstanceUuid, replacementCProductId,
+					discontinuedDateMonth, discontinuedDateDay,
+					discontinuedDateYear, discontinuedDateHour,
+					discontinuedDateMinute, serviceContext);
+			}
+		}
+
+		return addCPInstance(
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java:442]: use cpInstanceLocalService ref to make this method call

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-impl/src/main/java/com/liferay/headless/commerce/admin/catalog/internal/dto/v1_0/converter/SkuDTOConverter.java [hunk 8]
```diff
@@ -57,6 +57,22 @@ public Sku toDTO(DTOConverterContext dtoConverterContext) throws Exception {
 			{
 				cost = cpInstance.getCost();
 				depth = cpInstance.getDepth();
+
+				discontinued = cpInstance.isDiscontinued();
+
+				if (discontinued) {
+					CPInstance discontinuedCPInstance =
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-impl/src/main/java/com/liferay/headless/commerce/admin/catalog/internal/dto/v1_0/converter/SkuDTOConverter.java:64]: this should be called replacement Sku / somthing like that. 
I don't see the set in the DTO.
Also Brian want to have inline also for this object set with lambdas:
() -> {
    all your if + set the result in the dto
}

**Review by andrea-ale-sbarra — CHANGES_REQUESTED** [review]
See comments for details on changes.
Thanks!

**Comment by andrea-ale-sbarra** [comment]
sent here: https://github.com/liferay-commerce/liferay-portal/pull/2072

**Pull request closed without merging**